INTLY-3288 - UPS CSV

KevFan opened the pull request:
## Motivation
To allow Integreatly to consume the Unified Push Server Operator, a Cluster Service Version (CSV) would need to be generated. This pull request adds the generated csv that is to be used by integreatly.

Jira tickets include:
* https://issues.jboss.org/browse/INTLY-3288
* https://issues.jboss.org/browse/INTLY-3691

## What
Adds the Cluster Service Version (CSV) for the Unified Push Server Operator

## Why
This was added so the UPS operator can be used as part of the Integreatly Operator install

## How
This was done via the `operator-sdk olm-catalog gen-csv --csv-version=0.2.0` command for generating the base csv.

The csv was then further edited to include information on customresourcedefinitions, description, version, provider etc and the channel was updated to the `integreatly` channel to match the other csv in the [integreatly manifest repo](https://github.com/integr8ly/manifests).

The operator container image used in this CSV points to an image containing the `oauth-proxy` update - https://github.com/aerogear/unifiedpush-operator/pull/75 

## Verification Steps
* On a Openshift 4 cluster, create a new project 
* Create a operator source that points to my app registry:
```
apiVersion: operators.coreos.com/v1
kind: OperatorSource
metadata:
  name: integreatly-operators
  namespace: openshift-marketplace
spec:
  authorizationToken: {}
  displayName: Integreatly Operators
  endpoint: 'https://quay.io/cnr'
  publisher: Integreatly Publisher
  registryNamespace: kevfan
  type: appregistry
```
* Browse to the operator hub and verifiy that a Unifiedpush Operator with version `0.2.0` is present
* Verify that the `unifiedpush-operator` pod is created
* Kickstart the installation using the following `UnifiedPushServer.cr` as an example:
```
apiVersion: push.aerogear.org/v1alpha1
kind: UnifiedPushServer
metadata:
  name: example-unifiedpushserver
spec:
  useMessageBroker: false
```
* Verify the resources are created successully
* Visit the exposed ups route to ensure it works

**Note: As this is a copy of the csv that are going to be added to integreatly manifest repo, this should be merged when only the following pull requests are merged:**
* https://github.com/integr8ly/manifests/pull/70
* https://github.com/integr8ly/integreatly-operator/pull/127
## Checklist:

- [x] Code has been tested locally by PR requester

## Progress

- [x] Finished task

KevFan committed:
add csv

secondsun commented:
There is another PR open, https://github.com/aerogear/unifiedpush-operator/pull/71 that adds a new variant.  Will that need to update this as well?

grdryn commented:
@secondsun I guess that would depend on which of these is merged first :)

@KevFan is this complete now? When you're ready for it to be reviewed, just add me as a reviewer :+1:

KevFan commented:
@secondsun Yeah, I think the new variant should be added to the csv if this gets merged first 

@grdryn This should be good to for review as the other Integreatly PRs have been merged now. The csv contains the permissions recommended by @matskiv

grdryn commented:
@KevFan @matskiv have you managed to successfully run the operator & get it to deploy a UPS instance with the limited permissions defined here?

I was trying to update the role.yaml & role_binding.yaml to a similar set of permissions, but I hit errors, as described here: https://groups.google.com/d/msg/operator-framework/b36-lcGMxqU/afzTIOqBAQAJ

matskiv commented:
@grdryn Yes, it was working for me. I installed it from integreatly-operator, how are you setting up the operator and running it?


Also, two more permissions are required to avoid error messages (and in some case installation gets stuck due to this):
```
- verbs:
      - list
    apiGroups:
      - enmasse.io
    resources:
      - messagingusers
      - addressspaces
      - addresses
  - verbs:
      - list
    apiGroups:
      - user.enmasse.io
    resources:
      - messagingusers
```

grdryn commented:
@matskiv I'm just doing what we've always done on OpenShift 3, just with these modified permissions: https://github.com/grdryn/unifiedpush-operator/commit/43fc656b5833e7bdd4aba1481fb41d09021c6910

The problem I'm hitting seems to be that the Manager that's configured in main.go sets up a cache for all namespaces (which is needed for the pushapplication & variant CRs), but then the UnifiedPushServer controller complains that it can't set up watches for its primary & secondary resources at the cluster level, and just gets stuck in a loop there.

```
E1007 21:05:10.338585       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1alpha1.UnifiedPushServer: unifiedpushservers.push.aerogear.org is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list unifiedpushservers.push.aerogear.org at the cluster scope: no RBAC policy matched
E1007 21:05:10.341335       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.DeploymentConfig: deploymentconfigs.apps.openshift.io is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list deploymentconfigs.apps.openshift.io at the cluster scope: no RBAC policy matched
E1007 21:05:10.518825       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1beta1.CronJob: cronjobs.batch is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list cronjobs.batch at the cluster scope: no RBAC policy matched
E1007 21:05:10.571652       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list secrets at the cluster scope: no RBAC policy matched
E1007 21:05:10.706297       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.ServiceAccount: serviceaccounts is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list serviceaccounts at the cluster scope: no RBAC policy matched
E1007 21:05:10.718228       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.Route: routes.route.openshift.io is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list routes.route.openshift.io at the cluster scope: no RBAC policy matched
E1007 21:05:10.882025       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.PersistentVolumeClaim: persistentvolumeclaims is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list persistentvolumeclaims at the cluster scope: no RBAC policy matched
E1007 21:05:10.934222       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.ImageStream: imagestreams.image.openshift.io is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list imagestreams.image.openshift.io at the cluster scope: no RBAC policy matched
E1007 21:05:11.226903       1 reflector.go:134] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:126: Failed to list *v1.Service: services is forbidden: User "system:serviceaccount:unifiedpush:unifiedpush-operator" cannot list services at the cluster scope: no RBAC policy matched
```

Can you see anything obvious in my commit linked above that's missing? Does my understanding of the problem sound correct to you?

secondsun commented:
@KevFan Webpush got merged, can you add the new resources?

grdryn reviewed (dismissed):
Just updating the review status on this to be what we discussed last week: the permissions will need to be changed back.

KevFan committed:
use cluster permissions instead + add web push variant + add emasse permissions

laurafitzgerald commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -190,9 +200,6 @@ spec:
-                    fieldRef:
-                      fieldPath: metadata.annotations['olm.targetNamespaces']
                 - name: APP_NAMESPACES
                   value: unifiedpush-apps
```
Is it possible to override this value using OLM? We'd need to provide the MDC namespace in the case of integreatly?

KevFan replied:
I'm not too sure on this one. @matskiv may be able to better comment on this

matskiv replied:
Good catch. I don't know how to override that using OLM.
We should be able to just patch it from integreatly-operator, but that might cause other issues (e.g. OLM might revert it on upgrade?).

grdryn replied:
These annotations might help: https://docs.openshift.com/dedicated/4/applications/operators/olm-understanding-olm.html#olm-operatorgroups-csv-annotations_olm-understanding-olm

Alternatively, I'm happy to have this set to `unifiedpush-apps,mobile-developer-console,openshift-mobile-developer-console` by default (or even the latter two, and remove `unifiedpush-apps` altogether).

How does that sound @laurafitzgerald?

laurafitzgerald commented on deploy/olm-catalog/unifiedpush-operator/unifiedpush-operator.package.yaml:
```diff
@@ -0,0 +1,4 @@
+packageName: integreatly-unifiedpush
+channels:
+- currentCSV: unifiedpush-operator.v0.2.0
+  name: integreatly
```
I'm not exactly sure what channels do. Is integreatly the only one we want to include? Does that limit the use of the operator on OS4 upstream?

KevFan replied:
My understand are channels just points to specified csvs, to allow versions/upgrades depending on stability etc and it's possible to have multiple channels. So i don't think it should limit the use of the operator on OS4 upstream. Again @matskiv may be able to better comment on this, since I'm still quite new to this

matskiv replied:
@KevFan is right, it shouldn't limit us.
Also, for upstream we might just use a different organization for application repo, currently for integreatly we use: quay.io/application/integreatly/integreatly-unifiedpush
for upstream this would probably make more sense: quay.io/application/aerogear/unifiedpush
And we would not put "integreatly" channel into that repo.

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+      name: pushapplications.push.aerogear.org
+      version: v1alpha1
+      displayName: UPS Push Application
+      description: Represents an UPS Push Application Installation
```
```suggestion
      description: Represents a UPS Push Application
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+      name: androidvariants.push.aerogear.org
+      version: v1alpha1
+      displayName: Android Variant UPS
+      description: Represents an Android Variant UPS Installation
```
```suggestion
      description: Represents a UPS Android Variant
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+      name: iosvariants.push.aerogear.org
+      version: v1alpha1
+      displayName: IOS Variant UPS
+      description: Represents an IOS Variant UPS Installation
```
```suggestion
      description: Represents a UPS iOS Variant
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+    - kind: UnifiedPushServer
+      name: unifiedpushservers.push.aerogear.org
+      version: v1alpha1
+      displayName: Unified Push Server
```
```suggestion
      displayName: UnifiedPush Server
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+      name: unifiedpushservers.push.aerogear.org
+      version: v1alpha1
+      displayName: Unified Push Server
+      description: Represents an Unified Push Server Installation
```
```suggestion
      description: Represents an AeroGear UnifiedPush Server instance
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+    - kind: WebPushVariant
+      name: webpushvariants.push.aerogear.org
+      version: v1alpha1
+      displayName: Web Push Variant UPS
```
```suggestion
      displayName: WebPush Variant
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+    - kind: PushApplication
+      name: pushapplications.push.aerogear.org
+      version: v1alpha1
+      displayName: UPS Push Application
```
```suggestion
      displayName: Push Application
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+    - kind: AndroidVariant
+      name: androidvariants.push.aerogear.org
+      version: v1alpha1
+      displayName: Android Variant UPS
```
```suggestion
      displayName: Android Variant
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+    - kind: IOSVariant
+      name: iosvariants.push.aerogear.org
+      version: v1alpha1
+      displayName: IOS Variant UPS
```
```suggestion
      displayName: iOS Variant
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+      name: webpushvariants.push.aerogear.org
+      version: v1alpha1
+      displayName: Web Push Variant UPS
+      description: Represents an Web Push Variant Installation
```
```suggestion
      description: Represents a WebPush Variant
```

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+      displayName: Web Push Variant UPS
+      description: Represents an Web Push Variant Installation
+  description: The UnifiedPush Operator for Kubernetes provides an easy way to install and manage an AeroGear UnifiedPush Server on OpenShift.
+  displayName: Unifiedpush Operator
```
```suggestion
  displayName: UnifiedPush Operator
```

grdryn committed:
Update role.yaml for enmasse resources

This is just a copy of the permissions provided by the default
"enmasse.io:tenant-edit" ClusterRole.

grdryn committed:
Re-generate 0.2.0 CSV after rebase & role updates

grdryn committed:
Fix displayNames/descriptions in 0.2.0 CSV

grdryn commented on deploy/olm-catalog/unifiedpush-operator/0.2.0/unifiedpush-operator.v0.2.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,239 @@
+              - command:
+                - unifiedpush-operator
+                env:
+                - name: WATCH_NAMESPACE
```
By default this gets set to a reference to `metadata.annotations['olm.targetNamespaces']`. Did you have to change this manually @KevFan?

grdryn committed:
Set release image in operator.yaml

grdryn committed:
Generate 0.3.0 CSV from 0.2.0 CSV as a base

The command used to get this was:
operator-sdk olm-catalog gen-csv --csv-version=0.3.0 \
 --from-version=0.2.0 --update-crds --csv-channel=stable

The unifiedpush-operator.package.yaml ended up with a duplicate
channel entry though, but I think that's likely a bug in the
operator-sdk CLI:

https://github.com/operator-framework/operator-sdk/issues/1846

When we upgrade to a newer version, that should be fixed.

grdryn commented:
I've made a few changes here:
- Rebased
- Added missing permissions to role.yaml
- Regenerated the 0.2.0 after the rebase & role changes
- Generated the 0.3.0 CSV _from_ the 0.2.0 one (custom changes made for 0.2.0 are mostly copied over

There were some issues with the package.yaml, so I recreated it. I changed the "channel" name to "stable" too. I hope that's okay?

The command that I used to generate the 0.3.0 CSV was:
```
operator-sdk olm-catalog gen-csv --csv-version=0.3.0 \
     --from-version=0.2.0 --update-crds --csv-channel=stable
```

grdryn committed:
Add some optional CSV extras for operatorhub.io

Even though we won't be putting the operator on operatorhub.io yet,
this info can still be put here. I'm not sure if it will also surface
anywhere in OpenShift 4.

grdryn commented on deploy/olm-catalog/unifiedpush-operator/unifiedpush-operator.package.yaml:
```diff
@@ -0,0 +1,5 @@
+channels:
+- currentCSV: unifiedpush-operator.v0.3.0
+  name: stable
```
@KevFan @matskiv is there a need to have an 'integreatly' channel in here also?

matskiv replied:
I don't think so. We can have that in our copy in integr8ly/manifests repo.

grdryn committed:
🐛 Update APP_NAMESPACES to include mdc in CSVs

Not doing this in operator.yaml yet, since I think there's some dev
stuff in the Makefile that might need updating if we put more than one
namespace in there.
